Phosphor/src/edu/columbia/cs/psl/phosphor/TaintUtils.java

```diff
@@ -574,20 +574,20 @@ public static void arraycopyControlTrack(Object srcTaint, Object src, Object src
 					if (taints[i] == null)
 						taints[i] = ctrl.copyTag();
 					else {
-						taints[i].addDependency(ctrl.getTag());
+						taints[i] = Taint.addDependency(taints[i], ctrl.getTag());
 					}
 				}
 			}
 		} catch (ArrayIndexOutOfBoundsException ex) {
 			Taint t = null;
 			if (srcPosTaint != null) {
 				t = ((Taint) srcPosTaint).copy();
-				t.addDependency((Taint) destPosTaint);
-				t.addDependency((Taint) lengthTaint);
+				t = Taint.addDependency(t, (Taint) destPosTaint);
```

Comment:
It might be nice to have an addDependency that will add multiple dependencies at once, and then we can trivially cut down on the number of objects that we make and immediately throw away

Reply:
Right. What about this change to Taint.addDepndecy(Taint<T> oldTaint, Taint<T> d)? (which is the current method)

```
public static <T> Taint<T> addDependency(Taint<T> oldTaint, Taint<T>... d) {
    Taint<T> newTaint = Taint.createTaint(oldTaint);

    for (Taint<T> taint : d) {
        newTaint.addDependency(taint);
    }
    
    return newTaint;
}
```